src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt

```diff
@@ -284,23 +325,58 @@ class BackgroundLinkingIndexingService(
         return dataManager.markLinkingEntitiesAsIndexed(linkingIdsWithLastWrite)
     }
 
-    private fun lock(linkingIds: Collection<UUID>) {
-        linkingIds.forEach { linkingId ->
-            val existingExpiration = linkingIndexingLocks.putIfAbsent(
+
+    /**
+     * Checks if linking id is already "locked" in [linkingIndexingLocks] map.
+     * If it's not, it returns true, if it is, the function refreshes the expiration of the entry and returns false.
+     */
+    private fun lockOrRefresh(linkingId: UUID): Boolean {
+        val expiration = lockOrGet(linkingId)
+        // expiration should be >= now, otherwise entry should be evicted
+        if (expiration != null && Instant.now().toEpochMilli() <= expiration) {
+            // if it is about to expire (but not yet processed), we refresh the expiration/time to live
+            logger.info("Refreshing expiration for linking id {}", linkingId)
+            refreshExpiration(linkingId)
+        }
+
+        return expiration == null
+    }
+
+    /**
+     * @return Null if locked, expiration in millis otherwise.
+     */
+    private fun lockOrGet(linkingId: UUID): Long? {
+        return linkingIndexingLocks.putIfAbsent(
+                linkingId,
+                Instant.now().plusMillis(LINKING_INDEXING_TIMEOUT_MILLIS).toEpochMilli(),
+                LINKING_INDEXING_TIMEOUT_MILLIS,
+                TimeUnit.MILLISECONDS
+        )
+    }
+
+    private fun refreshExpiration(linkingId: UUID) {
+        try {
+            linkingIndexingLocks.lock(linkingId)
```

Comment:
Changed it to lock the whole `lockOrRefresh` process to avoid situations, where we start with expiration != null, but another process deletes it from the map and then the first process calls refresh and puts it back in